src/amp-story-player/page-scroller.js

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {throttle} from '../utils/rate-limit';
+
+/**
+ * Applies scroll and momentum to window by using touch events.
+ *
+ * It works using the following principles:
+ *
+ * On touchstart: (Re)set coordinates and timer. If user swiped while still
+ *                scrolling, increase multiplier for scrolling offset.
+ * On touchmove: Scroll 1:1 and calculate distance (deltaY) and acceleration of
+ *               the touch movement.
+ * On touchend: If meeting the momentum threshold, scroll by extra pixels
+ *              (scrolling offset) by a pre-set duration using an easing fn.
+ */
+export class PageScroller {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {number} */
+    this.cachedWinHeight_ = null;
+
+    /** @private {!Object} */
+    this.touchEventState_ = {
+      startY: 0,
+      lastDelta: 0,
+      touchStartTimeMs: null,
+      touchEndTimeMs: null,
+      touchMoveTimeMs: null,
+    };
+
+    /** @private {!Object} */
+    this.scrollState_ = {
+      startY: 0,
+      isRunning: false,
+      acceleration: 1,
+      speedLimit: 0.3,
+      startTimeMs: null,
+      maxTimeBetweenSwipesMs: 250,
+      moveTimeThresholdMs: 100,
+      durationMs: null,
+      meetsDeltaYThreshold: false,
+      deltaYThresholdPx: 5,
+      deltaY: null,
+      offsetThresholdPx: 30,
+      offsetPx: null,
+      multiplier: 1,
+    };
+  }
+
+  /**
+   * Reacts to onTouchStart events. (Re)sets coordinates and timer.
+   * Applies multiplier if criteria is met or resets it.
+   * @param {number} timeStamp
+   * @param {number} startY
+   */
+  onTouchStart(timeStamp, startY) {
+    this.touchEventState_.startY = startY;
+    this.touchEventState_.touchStartTimeMs = timeStamp;
+    this.scrollState_.startY = this.win_./*OK*/ scrollY;
+    this.cachedWinHeight_ =
+      this.cachedWinHeight_ || this.win_./*OK*/ innerHeight;
```

Comment:
If you change the size of the window this would not be updated at all, any reasons to cache and reuse the winHeight given that it's so easy to get?

Reply:
Yep good catch, reverted that change.